Feature/modular filehandling

klaseskilson opened the pull request:
Move file adding and downloading logic to the `Identity` model. The idea behind this is that the difference there might be between the different services, such as Dropbox, our own local storage and Drive _will be in which Identity the file belongs to_.

klaseskilson committed:
Add recursive folder adding

klaseskilson committed:
Use batch insert on folder adding, validate and add relationships

klaseskilson committed:
Load existing files with angular

klaseskilson committed:
Detect existing BruseFiles files on FilesController add-load

klaseskilson committed:
Keep track of existing files while browsing, too

klaseskilson committed:
Make file browsing clearer using BEM sass structure

It's not super pritty, just consider it as a concept.

klaseskilson committed:
Merge branch 'develop' into feature/dropbox-folders

Conflicts:
	app/assets/javascripts/controllers/FileCtrl.coffee

klaseskilson committed:
Merge branch 'develop' of github.com:tkomstadius/bruse into feature/dropbox-folders

Conflicts:
	app/assets/javascripts/factories/FileHandler.coffee
	app/controllers/files_controller.rb
	app/models/bruse_file.rb

klaseskilson committed:
Add more test cases to FilesController

klaseskilson committed:
Fix font awesome issue

klaseskilson committed:
Add some documentation to FileCtrl

klaseskilson committed:
Merge branch 'develop' into feature/dropbox-folders

klaseskilson committed:
Make sure the current identity belongs to current_user

klaseskilson committed:
Add `.id` to FilesController test user, closes #12

klaseskilson committed:
Create file browsing method in identity

klaseskilson committed:
Add single file on identity model instead of in controller

klaseskilson committed:
Add recursive folder add to identity

klaseskilson committed:
Add more BruseFile and Identity tests

klaseskilson committed:
Do not return an array with arrays from files#index

klaseskilson committed:
Use modular functions for browsing and folder detection

klaseskilson committed:
Fix problem with files from added folders not appearing as added

The issue was that files was pushed to the $scope.bruse_file array
as an array, resulting in an array of arrays. This made it
impossible for the findFile method to search through it, since it
only search through one level.

klaseskilson committed:
Move download things to Identity

klaseskilson commented on Gemfile:
```diff
@@ -35,6 +37,8 @@ gem 'font-awesome-rails'
 # sass mixins that makes you happy
 gem 'bourbon'
+# normalize!
+gem 'normalize-rails'
```
Normalize resets the css, so that any Browser-specific special default css styles is removed.

tkomstadius commented:
You can only add folders, not remove them

klaseskilson commented:
Yes, that is correct. Working on it!

ronjagrosz commented:
Adding a folder works, but the minus sign does not appear for the individual files in the folder. Pressing the plus sign gives the feedback that the file already exists. :(

klaseskilson commented:
@tkomstadius, do you have the same issue as @Fluffpannkaka?

tkomstadius commented:
No, I don't think so. I still haven't got a minus for the folder but if i add then collapse all the individual files have a minus. But when i remove files, add the folder again the removed files isn't added again.

klaseskilson committed:
Add folder deletion

All files on the identity that has a foreign_ref beginning with
the folder name gets deleted.

ronjagrosz commented:
It works now! :D only seems to be a refresh issue. When adding the folder, expanding, the minus signs are there. If I then click the minus button on the folder the feedback in the console says that they are removed, but the signs do not return to plus on the individual files, they still have(-). Only for the folder, which now has (+). Maybe do some angular for refreshing the icons(?)

tkomstadius commented:
I can't remove files anymore. They just gets added multiple times instead

tkomstadius commented:
I also get a lot of "could not unsave file" (404) when I try to remove them.

ronjagrosz commented:
Ah no. I can't add files either.

klaseskilson committed:
Make sure files can be added and deleted

klaseskilson commented:
Found a couple of more issues, too. When removing files, they still exist in `$scope.bruse_files`. When removing folders, the files in that folder still exist in `$scope.bruse_file` _and_ on the file remote file object in `$scope.files`.

The icons totally depend on the remote file object.

klaseskilson commented:
This could possibly be resolved using the `collect` method in `FileHandler`, since that's an easy way to make sure we have all existing files loaded and then the [`findFile` method in `FileCtrl`](https://github.com/tkomstadius/bruse/blob/5ca4716301cd43a0b6d1aacf15c09d844dc296d9/app/assets/javascripts/controllers/FileCtrl.coffee#L84). 

Much like we [do when the page loads](https://github.com/tkomstadius/bruse/blob/5ca4716301cd43a0b6d1aacf15c09d844dc296d9/app/assets/javascripts/controllers/FileCtrl.coffee#L9-L22) and when we [expand a folder](https://github.com/tkomstadius/bruse/blob/5ca4716301cd43a0b6d1aacf15c09d844dc296d9/app/assets/javascripts/controllers/FileCtrl.coffee#L38-L39).

klaseskilson commented:
I feel this part of the system is consuming too much development time. Most of the features in this PR is about importing external files into Bruse, and not handling the actual BruseFiles. Therefore, I think we can merge this PR into develop, and then branch out to implement batch (multiple) handling of the existing BruseFiles. I suggest that we also try to implement a way to separate the already existing files and the newly added files to show the user what's happening.

Does that sound OK?

tkomstadius commented:
Yes, if that will help us move forward I think that that sounds like a good plan. Then I feel this is ready for merge.

danielronnkvist commented:
I think it looks good!
